tensorflow_probability/python/mcmc/dual_averaging_step_size_adaptation.py

```diff
@@ -0,0 +1,484 @@
+        # non-structure-broadcasted case.
+        #
+        # It turns out we can compute the reduction dimensions for both cases
+        # uniformly by taking the rank of any state part. This obviously works
+        # in the second case (where all state ranks are the same). In the first
+        # case, all state parts have the rank L + D_i + B, where L is the rank
+        # of log_accept_prob, D_i is the non-shared dimensions amongst all
+        # states, and B are the shared dimensions of all the states, which are
+        # equal to the step size. When we subtract B, we will always get a
+        # number >= L, which means we'll get the full reduction we want.
+        (step_size_part, state_part, error_sum_part,
+         log_averaging_step_part, shrinkage_target_part) = parts
+        num_reduce_dims = tf.minimum(
+            log_accept_prob_rank,
+            tf.rank(state_part) - tf.rank(step_size_part))
+        reduced_log_accept_prob = _reduce_logmeanexp(log_accept_prob,
+                                                     tf.range(num_reduce_dims))
+
+        # reduced_log_accept_prob must broadcast into step_size_part on the
+        # left, so we do an additional reduction over dimensions where their
+        # shapes differ.
+        reduce_indices = _get_differing_dims(reduced_log_accept_prob,
+                                             step_size_part)
+        reduced_log_accept_prob = _reduce_logmeanexp(
+            reduced_log_accept_prob, reduce_indices, keepdims=True)
+        new_error_sum = (error_sum_part +
+                         target_accept_prob -
+                         tf.math.exp(reduced_log_accept_prob))
+
+        initial_t = previous_kernel_results.initial_t
+        step = tf.cast(previous_kernel_results.step, initial_t.dtype) + 1.
+        soft_t = initial_t + step
+
+        new_log_step = (shrinkage_target_part - (
+            tf.reshape(new_error_sum, shrinkage_target_part.shape) *
+            tf.math.sqrt(step)) / (
+                soft_t * previous_kernel_results.exploration_shrinkage))
+
+        eta = tf.math.pow(step, -previous_kernel_results.decay_rate)
+        new_log_averaging_step = (eta * new_log_step +
+                                  (1 - eta) * log_averaging_step_part)
+
+        # - If still adapting, return an exploring step size,
+        # - If just finished, return the averaging step size
+        # - Otherwise, do not update
+        new_step_size_parts.append(
+            tf.where(
+                previous_kernel_results.step < self.num_adaptation_steps,
+                tf.math.exp(new_log_step),
+                tf.where(
+                    previous_kernel_results.step > self.num_adaptation_steps,
+                    step_size_part,
+                    tf.math.exp(new_log_averaging_step))))
+        new_log_averaging_step_parts.append(
+            tf.where(
+                previous_kernel_results.step > self.num_adaptation_steps,
```

Comment:
Let me know if I misunderstood above, since I bet both should have the same answer.

I think keeping this running after fixing the step size will be somewhat misleading, since you do not have the accumulating error informing the step size being used -- just checking acceptance probabilities should give you the same information.